### PR TITLE
fix: div and rem by negative zero

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/521 and https://github.com/apache/datafusion-comet/issues/665.

## What changes are included in this PR?
This PR normalize the right hand of div and rem

## How are these changes tested?
Test updated

### CODE REVIEW

File: docs/source/contributor-guide/development.md [hunk 1]
```diff
@@ -91,13 +91,15 @@ Comet has a plan stability testing framework that can be used to test the stabil
 The plan stability testing framework is located in the `spark` module and can be run using the following command:
 
 ```sh
-./mvnw -pl spark -Dsuites="org.apache.spark.sql.comet.CometTPCDSV1_4_PlanStabilitySuite" test
+./mvnw -pl spark -Dsuites="org.apache.spark.sql.comet.CometTPCDSV1_4_PlanStabilitySuite" -nsu test
+./mvnw -pl spark -Dsuites="org.apache.spark.sql.comet.CometTPCDSV1_4_PlanStabilitySuite" -Pspark-3.5 -nsu test
```
Comment (mbutrovich) [inline docs/source/contributor-guide/development.md:95]: Since you added these steps to the testing commands, does it makes sense to duplicate them below for the generation commands?

**Comment by kazuyukitanimura** [comment]
cc @andygrove @comphead @huaxingao @mbutrovich @parthchandra @viirya

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 2]
```diff
@@ -873,12 +873,18 @@ class CometSparkSessionExtensions
...
+          Divide(left, normalizeNaNAndZero(right), evalMode)
+        case Remainder(left, right, evalMode) =>
+          Remainder(left, normalizeNaNAndZero(right), evalMode)
       }
     }
 
     def normalizeNaNAndZero(expr: Expression): Expression = {
       expr match {
         case _: KnownFloatingPointNormalized => expr
+        case FloatLiteral(f) if !f.equals(-0.0f) => expr
```
Comment (mbutrovich) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:886]: Just clarifying: this is saying that as long as the literal isn't -0.0, we don't have to normalize it? And it doesn't look like there's any way to define a literal NaN, so we don't handle that case?

**Review by mbutrovich — COMMENTED** [review]
Just comments and questions from me.

File: docs/source/contributor-guide/development.md (hunk 1 above)
Comment (kazuyukitanimura) [reply docs/source/contributor-guide/development.md:95]: Thanks, updated.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 2 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:886]: > Just clarifying: this is saying that as long as the literal isn't -0.0, we don't have to normalize it? 
Exactly

> And it doesn't look like there's any way to define a literal NaN, so we don't handle that case?
We can create a literal NaN, but Spark `NormalizeNaNAndZero` is not doing much for literal NaN. I think NaN normalization matters only when some expression results are NaN https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NormalizeFloatingNumbers.scala#L194

File: docs/source/contributor-guide/development.md [hunk 3]
```diff
@@ -91,27 +91,31 @@ Comet has a plan stability testing framework that can be used to test the stabil
 The plan stability testing framework is located in the `spark` module and can be run using the following command:
 
 ```sh
-./mvnw -pl spark -Dsuites="org.apache.spark.sql.comet.CometTPCDSV1_4_PlanStabilitySuite" test
+./mvnw -pl spark -Dsuites="org.apache.spark.sql.comet.CometTPCDSV1_4_PlanStabilitySuite" -nsu test
```
Comment (viirya) [inline docs/source/contributor-guide/development.md:94]: Why add `-nsu`?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 4]
```diff
@@ -2636,7 +2636,11 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
 
   def nullIfWhenPrimitive(expression: Expression): Expression = if (isPrimitive(expression)) {
     val zero = Literal.default(expression.dataType)
-    If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+    expression match {
+      case _: Literal if expression != zero => expression
```
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2640]: Is this related?

File: docs/source/contributor-guide/development.md (hunk 3 above)
Comment (kazuyukitanimura) [reply docs/source/contributor-guide/development.md:94]: It helps for faster turnaround

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 4 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2640]: Indirectly related, this will avoid to put `EqualTo` for non zero literals for divisions

**Review by viirya — APPROVED** [review]

**Comment by kazuyukitanimura** [comment]
Closes 585

**Comment by kazuyukitanimura** [comment]
Merged Thank you @mbutrovich @viirya
